### PR TITLE
doc: add wraping for unit tests

### PR DESCRIPTION
**Pull request by dabekjakub**
Added guide for symbol wraping in unit tests.

Signed-off-by: Jakub Dabek <jakub.dabek@linux.intel.com>

### CODE REVIEW

**Comment by lgirdwood** [comment]
@deb-intel ping. Do you approve ?

File: developer_guides/unit_tests.rst [hunk 1]
```diff
@@ -84,6 +84,30 @@ Preparing cmocka package
 
       ./configure --with-cmocka-prefix=/home/<you>/cminstall ...
 
+Wraping objects for unit tests
+******************************
+
+If at any point should You need to mock any symbol it can be done by simply defining it in unit test and only including .h files. However there can exist corner cases in which function that should be mocked can't be done in this way. The cases are - Static functions in headers(those most probably are inline short functions and don't have to be mocked) or static functions that are in the same file as tested functionality and are exceedingly large so they can't be tested as one functionality. Whatever the reason mocking of those symbols can be done by using --wrap linker functionality. To wrap the symbol follow these steps:
```
Comment (intelkevinputnam) [inline developer_guides/unit_tests.rst]: Line 90 - "If you need to mock a symbol, define it in a unit test and include the .h file. There are 2 cases where this isn't possible:

* Static functions in headers...
* Static functions that are ...

Whatever the reason ...

File: developer_guides/unit_tests.rst [hunk 2]
```diff
@@ -84,6 +84,30 @@ Preparing cmocka package
...
+Wraping objects for unit tests
+******************************
+
+If at any point should You need to mock any symbol it can be done by simply defining it in unit test and only including .h files. However there can exist corner cases in which function that should be mocked can't be done in this way. The cases are - Static functions in headers(those most probably are inline short functions and don't have to be mocked) or static functions that are in the same file as tested functionality and are exceedingly large so they can't be tested as one functionality. Whatever the reason mocking of those symbols can be done by using --wrap linker functionality. To wrap the symbol follow these steps:
+
+#. Create mocked symbol named __wrap_symbol_name
+
+#. Pass instruction for the linker -Wl, --wrap=symbol_name during compilation. Now every symbol call to symbol_name will call __wrap_symbol_name.
+
+Passing instruction to linker in SOF UT enviroment can be done by use of CFLAGS. However this is a subject to change since linker flags should be passed in separate variable in makefiles.
```
Comment (intelkevinputnam) [inline developer_guides/unit_tests.rst]: Line 96 - "Instructions can be passed to the linker in the SOF UT environment using CFLAGS, however they should be passed in separate variables in the makefile."

**Comment by dabekjakub** [comment]
Changed phrasing and formating as per @intelkevinputnam comments.

File: developer_guides/unit_tests.rst [hunk 3]
```diff
@@ -84,6 +84,43 @@ Preparing cmocka package
...
+If you need to mock a symbol, define it in a unit test and include the .h file. 
+There are 2 cases where this isn't possible:
+
+*	Static functions in headers(those most probably are inline short functions
+	and don't have to be mocked)
+
+*	Static functions that are in the same file as tested functionality and are
+	exceedingly large so they can't be tested as one functionality. 
+
+Whatever the reason mocking of those symbols can be done by using --wrap linker
```
Comment (intelkevinputnam) [inline developer_guides/unit_tests.rst]: Line 99: Add a comma "Whatever the reason, "

File: developer_guides/unit_tests.rst [hunk 4]
```diff
@@ -84,6 +84,43 @@ Preparing cmocka package
...
+#. Pass instruction for the linker -Wl, --wrap=symbol_name during compilation.
+
+Now every symbol call to symbol_name will call __wrap_symbol_name.
+
+Instructions can be passed to the linker in the SOF UT environment using
+CFLAGS, however they should be passed in separate variables in the makefile.
+
+Example:
+
+	.. code-block:: bash
```
Comment (intelkevinputnam) [inline developer_guides/unit_tests.rst]: Line 113: .. code-block:: bash should not be indented just the lines following it.

File: developer_guides/unit_tests.rst [hunk 5]
```diff
@@ -84,6 +84,43 @@ Preparing cmocka package
...
+
+	  # some tests before ...
+          check_PROGRAMS += pipeline_connect_upstream
+          pipeline_connect_upstream_SOURCES = ../../src/audio/pipeline.c src/audio/pipeline/ pipeline_mocks.c src/audio/pipeline/pipeline_connect_upstream.c src/audio/pipeline/pipeline_mocks_rzalloc.c
+          pipeline_connect_upstream_CFLAGS = -Wl, --wrap=symbol_name
+
+Full information about wraping can be found here:
```
Comment (intelkevinputnam) [inline developer_guides/unit_tests.rst]: Line 120: change to "wrapping"

**Comment by dabekjakub** [comment]
Changed in accordance to comments , and aligned rest of the file to the indent standard.

File: developer_guides/unit_tests.rst [hunk 6]
```diff
@@ -56,7 +56,7 @@ Preparing cmocka package
 
 1. Build cmocka with static library on:
 
-   .. code-block:: bash
+.. code-block:: bash
```
Comment (dbkinder) [inline developer_guides/unit_tests.rst]: @deb-intel outdenting all these code blocks will cause the numbered list to reset to 1 for the next item.  The indented code block directive that was there was correct since the code block is part of the list item material.

File: developer_guides/unit_tests.rst [hunk 7]
```diff
@@ -80,10 +80,47 @@ Preparing cmocka package
 
 #. Use the target location for cmocka files when invoking *configure* script:
 
-   .. code-block:: bash
+.. code-block:: bash
 
       ./configure --with-cmocka-prefix=/home/<you>/cminstall ...
 
+Wrapping objects for unit tests
+******************************
```
Comment (dbkinder) [inline developer_guides/unit_tests.rst]: @deb-intel this underline is too short

File: developer_guides/unit_tests.rst [hunk 8]
```diff
@@ -80,10 +80,47 @@ Preparing cmocka package
...
+Instructions can be passed to the linker in the SOF UT environment using
+CFLAGS, however they should be passed in separate variables in the makefile.
+
+Example:
+
+.. code-block:: bash
+ 
+	  # some tests before ...
+          check_PROGRAMS += pipeline_connect_upstream
+          pipeline_connect_upstream_SOURCES = ../../src/audio/pipeline.c src/audio/pipeline/ pipeline_mocks.c src/audio/pipeline/pipeline_connect_upstream.c src/audio/pipeline/pipeline_mocks_rzalloc.c
```
Comment (dbkinder) [inline developer_guides/unit_tests.rst]: better to break the long command lines and end with a continuation character ``\``

File: developer_guides/unit_tests.rst (hunk 7 above)
Comment (jajanusz) [reply developer_guides/unit_tests.rst]: fixed in #29 https://github.com/thesofproject/sof-docs/pull/29/commits/4207d17ac70bb98c6d9c1f9faa6f62774d58b4e4
